docx/from/docxtotei.xsl

```diff
@@ -1,713 +1,743 @@
+      <xsl:for-each-group select="w:sdt|w:p|w:tbl"
+        group-starting-with="w:p[tei:isFirstlevel-heading(.)]">
+
+        <xsl:choose>
+
+          <!-- We are dealing with a first level section, we now have
+         to further divide the section into subsections that we can then
+         finally work on -->
+
+    <xsl:when test="tei:is-heading(.)">
+      <xsl:call-template name="group-by-section"/>
+    </xsl:when>
+
+    <xsl:when test="tei:is-front(.)">
+      <front>
+        <xsl:apply-templates select="." mode="inSectionGroup"/>
+      </front>
+    </xsl:when>
+
+    <!-- We have found some loose paragraphs. These are most probably
+         front matter paragraps. We can simply convert them without further
+         trying to split them up into sub sections. -->
+    <xsl:otherwise>
+      <xsl:apply-templates select="." mode="inSectionGroup"/>
+    </xsl:otherwise>
+  </xsl:choose>
+</xsl:for-each-group>
+
+<!-- I have no idea why I need this, but I apparently do. 
+     //TODO: find out what is going on-->
+      <xsl:apply-templates select="w:sectPr" mode="paragraph"/>
+    </xsl:template>
+    <doc xmlns="http://www.oxygenxml.com/ns/doc/xsl">
+      <desc>
+        <p>Bookmarks in section mode</p>
+        <p>
+          There are certain elements that we don't really care about, but that
+          force us to regroup everything from the next sibling on.
+
+          @see grouping in construction of headline outline.
+        </p>
+      </desc>
+    </doc>
+    <xsl:template match="w:bookmarkStart|w:bookmarkEnd"
+      mode="inSectionGroup">
+      <xsl:for-each-group select="current-group() except ." group-adjacent="1">
+        <xsl:apply-templates select="." mode="inSectionGroup"/>
+      </xsl:for-each-group>
+    </xsl:template>
+
+
+    <doc xmlns="http://www.oxygenxml.com/ns/doc/xsl">
+      <desc>
+        <p>Bookmarks in normal mode</p>
+        <p>Copy bookmarks for processing in pass 2</p>
+      </desc>
+    </doc>
+    <xsl:template match="w:bookmarkStart|w:bookmarkEnd" >
+      <xsl:choose>
+        <xsl:when test="@w:name='_GoBack'"/>
+        <xsl:when test="starts-with(@w:name,'_')">
+          <xsl:copy-of select="."/>
+        </xsl:when>
+        <xsl:when test="self::w:bookmarkEnd"/>
+        <xsl:otherwise>
+          <ANCHOR>
+            <xsl:attribute name="xml:id" select="@w:name"/>
+          </ANCHOR>
+        </xsl:otherwise>
+      </xsl:choose>
+    </xsl:template>
+
+    <xsl:template name="removeNestedTables" match="w:txbxContent/w:tbl" mode="passOneHalf">
+      Could not process table here.  Table was removed.
+    </xsl:template>
+
+    <!-- We cannot process pict directly, but we can try to process textboxes inside them. -->
+    <xsl:template name="unnestFromPict" match="w:pict" mode="passOneHalf">
+      <xsl:apply-templates  select="v:shape/v:textbox/w:txbxContent/*" mode="passOneHalf" />
+    </xsl:template>
+
+
+    <xsl:template match="/ | @* | node()" mode="passOneHalf">
```

Comment:
I'll chat or Screenhero any time. :smile:

I'll give it a try here though.  Templates are applied from the most specific match to the least specific match. But once a match is made, no future templates are applied.  

This is what caused the problem we saw where the raw Word XML was making it in to the final document.  We grabbed all of the content we wanted, but since we didn't do it in an `apply-templates` call, all processing stopped.

That's also why we need the identity transform in this pass.  If a node doesn't match any of the selectors in a pass, it won't be selected and passed on.  The identity selection is the least specific call, so it runs last and makes sure we get the whole document.

And yes, if you call `apply-templates mode="foo"` from an `apply-templates mode="foo"` that template will get applied again to all content that matches the selectors.  This is what we want to happen, but can lead to some pretty deep recursion when applied improperly.  It's a consideration for development, and I did see a stackoverflow once or twice, but once the selectors are set up properly we should be good.